Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.cpp

```diff
@@ -0,0 +1,243 @@
+    m_icon_view = m_icon_view_container->add<GUI::IconView>();
+    m_icon_view->set_always_wrap_labels(true);
+    m_icon_view->set_model(m_model);
+    m_icon_view->set_model_column(ProjectTemplatesModel::Column::Name);
+    m_icon_view->on_selection_change = [&]() {
+        update_dialog();
+    };
+    m_icon_view->on_activation = [&]() {
+        if (m_input_valid)
+            do_create_project();
+    };
+
+    m_description_label = *main_widget.find_descendant_of_type_named<GUI::Label>("description_label");
+    m_name_input = *main_widget.find_descendant_of_type_named<GUI::TextBox>("name_input");
+    m_name_input->on_change = [&]() {
+        update_dialog();
+    };
+    m_name_input->on_return_pressed = [&]() {
+        if (m_input_valid)
+            do_create_project();
+    };
+    m_create_in_input = *main_widget.find_descendant_of_type_named<GUI::TextBox>("create_in_input");
+    m_create_in_input->on_change = [&]() {
+        update_dialog();
+    };
+    m_create_in_input->on_return_pressed = [&]() {
+        if (m_input_valid)
+            do_create_project();
+    };
+    m_full_path_label = *main_widget.find_descendant_of_type_named<GUI::Label>("full_path_label");
+
+    m_ok_button = *main_widget.find_descendant_of_type_named<GUI::Button>("ok_button");
+    m_ok_button->on_click = [this](auto) {
+        do_create_project();
+    };
+
+    m_cancel_button = *main_widget.find_descendant_of_type_named<GUI::Button>("cancel_button");
+    m_cancel_button->on_click = [this](auto) {
+        done(ExecResult::ExecCancel);
+    };
+
+    m_browse_button = *find_descendant_of_type_named<GUI::Button>("browse_button");
+    m_browse_button->on_click = [this](auto) {
+        Optional<String> path = GUI::FilePicker::get_open_filepath(this);
+        if (path.has_value())
+            m_create_in_input->set_text(path.value().view());
+    };
+}
+
+NewProjectDialog::~NewProjectDialog()
+{
+}
+
+RefPtr<ProjectTemplate> NewProjectDialog::selected_template()
+{
+    if (m_icon_view->selection().is_empty()) {
+        return {};
+    }
+
+    auto project_template = m_model->template_for_index(m_icon_view->selection().first());
+    ASSERT(!project_template.is_null());
+
+    return project_template;
+}
+
+void NewProjectDialog::update_dialog()
+{
+    auto project_template = selected_template();
+    m_input_valid = true;
+
+    if (project_template) {
+        m_description_label->set_text(project_template->description());
+    } else {
+        m_description_label->set_text("Select a project template to continue.");
+        m_input_valid = false;
+    }
+
+    auto maybe_project_path = get_project_full_path();
+
+    if (maybe_project_path.has_value()) {
+        m_full_path_label->set_text(maybe_project_path.value());
+    } else {
+        m_full_path_label->set_text("Invalid name or creation directory.");
+        m_input_valid = false;
+    }
```

Comment:
Same as for the project template, having the project path be an `Option<String>` instead of a `Result<String, String>` limits the message we can show to the user ("Invalid" isn't as useful as we could be, given that the `get_project_full_path()` method can fail in 6 different ways both by itself or in `get_available_project_name()`).

Reply:
Referring to the previous conversation, it seems I'll have to pass on this for now until I find out how to specify an error value for `Result`s with the same value and error type 😄 Shouldn't be too much of a problem for now.

Reply:
Wow, that's right ! I hadn't even considered the possibility thiat it couldn't be done, and that seems like an unfortunate limitation of the `Result` type.
I don't have any issues with skipping it for this PR. It's just a small usability thing that can be improved later once the `Result<T, T>` thing is possible!

Reply:
An alternative would be to have a :
```cpp
struct TemplateCreationError {
    TemplateCreationError(String message)
        : message(message)
    {
    }
    String message;
}
```
and use it as the error type. This way, you could write `return TemplateCreationError("My message");`

Reply:
Hmm... I've managed to pass along the template creation error with a `Result<void, String>` so far, do you think it would be more correct defining a specific error struct for this circumstance, even though only a String is being passed at the moment?

Reply:
If you hold on just a few minutes I'll push my new changes up (unsquashed) :)

Reply:
I don't actually know if creating a specific struct for errors is good practice in C++.
@awesomekling thoughts about using `Result<void, String>` vs `Result<void, TemplateCreationError>`, where `TemplateCreationError` is a wrapper struct around a string ?